File: src/dotnet/cdk/Services/Inventory.Acl/InventoryAclService.cs
using System.Collections.Generic;
using Amazon.CDK;
using Amazon.CDK.AWS.Events;
using Amazon.CDK.AWS.Events.Targets;
using Amazon.CDK.AWS.Lambda.EventSources;
using Amazon.CDK.AWS.SecretsManager;
using Amazon.CDK.AWS.SNS;
using Amazon.CDK.AWS.SSM;
using Constructs;
using ServerlessGettingStarted.CDK.Constructs;

namespace ServerlessGettingStarted.CDK.Services.Inventory.Acl;

public record InventoryAclServiceProps(SharedProps Shared, ISecret DdApiKeySecret, IEventBus SharedEventBus);

public class InventoryAclService : Construct
{
    public InventoryAclService(Construct scope, string id, InventoryAclServiceProps props) : base(scope, id)
    {
        var productAddedHandlerQueue = new ResilientQueue(this, "ProductCreatedEventQueue",
            new ResilientQueueProps("Inventory-ProductCreated", props.Shared.Env));
        var productUpdatedHandlerQueue = new ResilientQueue(this, "ProductUpdatedEventQueue",
            new ResilientQueueProps("Inventory-ProductUpdated", props.Shared.Env));
        
        var newProductAddedTopic = new Topic(this, "NewProductAdded", new TopicProps()
        {
            TopicName = $"DotnetNewProductAdded-{props.Shared.Env}"
        });
        
        var apiEnvironmentVariables = new Dictionary<string, string>(2)
        {
            { "PRODUCT_ADDED_TOPIC_ARN", newProductAddedTopic.TopicArn },
        };
        
        var handleProductCreatedFunction = new InstrumentedFunction(this, "HandleProductCreatedFunction",
            new FunctionProps(props.Shared,"HandleProductCreated", "../src/Inventory.Acl/Inventory.Acl.Adapters/",
                "Inventory.Acl.Adapters::Inventory.Acl.Adapters.HandlerFunctions_HandleCreated_Generated::HandleCreated", apiEnvironmentVariables, props.DdApiKeySecret));
        newProductAddedTopic.GrantPublish(handleProductCreatedFunction.Function);
        handleProductCreatedFunction.Function.AddEventSource(new SqsEventSource(productAddedHandlerQueue.Queue, new SqsEventSourceProps()
        {
            ReportBatchItemFailures = true,
            BatchSize = 10,
            MaxBatchingWindow = Duration.Seconds(10)
        }));
        
        var handleProductUpdatedFunction = new InstrumentedFunction(this, "HandleProductUpdatedFunction",
            new FunctionProps(props.Shared,"HandleProductUpdated", "../src/Inventory.Acl/Inventory.Acl.Adapters/",
                "Inventory.Acl.Adapters::Inventory.Acl.Adapters.HandlerFunctions_HandleCreated_Generated::HandleUpdated", apiEnvironmentVariables, props.DdApiKeySecret));
        newProductAddedTopic.GrantPublish(handleProductUpdatedFunction.Function);
        handleProductUpdatedFunction.Function.AddEventSource(new SqsEventSource(productUpdatedHandlerQueue.Queue, new SqsEventSourceProps()
        {
            ReportBatchItemFailures = true,
            BatchSize = 10,
            MaxBatchingWindow = Duration.Seconds(10)
        }));

        var productCreatedRule = new Rule(this, "InventoryProductCreatedRule", new RuleProps()
        {
            EventBus = props.SharedEventBus,
        });
        productCreatedRule.AddEventPattern(new EventPattern()
        {
            DetailType = ["product.productCreated.v1"],
            Source = [$"{props.Shared.Env}.products"]
        });
        productCreatedRule.AddTarget(new SqsQueue(productAddedHandlerQueue.Queue));
        var productUpdatedRule = new Rule(this, "InventoryProductUpdatedRule", new RuleProps()
        {
            EventBus = props.SharedEventBus,
        });
        productUpdatedRule.AddEventPattern(new EventPattern()
        {
            DetailType = ["product.productUpdated.v1"],
            Source = [$"{props.Shared.Env}.products"]
        });
        productUpdatedRule.AddTarget(new SqsQueue(productUpdatedHandlerQueue.Queue));
        
        var pricingUpdatedTopic = new StringParameter(this, "PricingUpdatedTopicArn",
            new StringParameterProps()
            {
                ParameterName = "/dotnet/inventory-acl/new-product-added-topic",
                StringValue = newProductAddedTopic.TopicArn
            });
    }
}

File: src/go/cdk/inventoryAcl/inventoryAclService.go
package inventoryacl

import (
	sharedprops "cdk/shared"
	sharedconstructs "cdk/sharedConstructs"
	"fmt"

	"github.com/aws/aws-cdk-go/awscdk/v2/awsevents"
	"github.com/aws/aws-cdk-go/awscdk/v2/awseventstargets"
	"github.com/aws/aws-cdk-go/awscdk/v2/awslambdaeventsources"
	"github.com/aws/aws-cdk-go/awscdk/v2/awssns"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsssm"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type InventoryAclServiceProps struct {
	SharedProps    sharedprops.SharedProps
	SharedEventBus awsevents.IEventBus
}

func NewInventoryAclService(scope constructs.Construct, id string, props *InventoryAclServiceProps) {
	inventoryProductAddedTopic := awssns.NewTopic(scope, jsii.String("GoInventoryProductAddedTopic"), &awssns.TopicProps{
		TopicName: jsii.Sprintf("GoInventoryProductAddedTopic-%s", props.SharedProps.Env),
	})

	productCreatedPublicEventQueue := sharedconstructs.NewResiliantQueue(scope, "GoProductCreatedPublicEventQueue", &sharedconstructs.ResiliantQueueProps{
		SharedProps: props.SharedProps,
		QueueName:   "GoProductCreatedPublicEventQueue",
	})

	environmentVariables := make(map[string]*string)
	environmentVariables["PRODUCT_ADDED_TOPIC_ARN"] = jsii.String(*inventoryProductAddedTopic.TopicArn())

	publicEventPublisherFunction := sharedconstructs.NewInstrumentedFunction(scope, "InventoryAcl", &sharedconstructs.InstrumentedFunctionProps{
		SharedProps:          props.SharedProps,
		FunctionName:         "GoInventoryAcl",
		Entry:                "../src/inventory-acl/product-created-public-event-handler/",
		EnvironmentVariables: environmentVariables,
	})

	inventoryProductAddedTopic.GrantPublish(publicEventPublisherFunction.Function)

	publicEventPublisherFunction.Function.AddEventSource(awslambdaeventsources.NewSqsEventSource(productCreatedPublicEventQueue.Queue, &awslambdaeventsources.SqsEventSourceProps{
		ReportBatchItemFailures: jsii.Bool(true),
	}))

	productCreatedRule := awsevents.NewRule(scope, jsii.String("Inventory-OrderCreated"), &awsevents.RuleProps{
		EventBus: props.SharedEventBus,
	})

	productCreatedRule.AddEventPattern(&awsevents.EventPattern{
		DetailType: jsii.Strings("product.productCreated.v1"),
		Source:     jsii.Strings(fmt.Sprintf("%s.products", props.SharedProps.Env)),
	})
	productCreatedRule.AddTarget(awseventstargets.NewSqsQueue(productCreatedPublicEventQueue.Queue, &awseventstargets.SqsQueueProps{}))

	awsssm.NewStringParameter(scope, jsii.String("GoInventoryProductAddedTopicParam"), &awsssm.StringParameterProps{
		ParameterName: jsii.String("/go/inventory/product-added-topic"),
		StringValue:   inventoryProductAddedTopic.TopicArn(),
	})

	awsssm.NewStringParameter(scope, jsii.String("GoInventoryProductAddedTopicNameParam"), &awsssm.StringParameterProps{
		ParameterName: jsii.String("/go/inventory/product-added-topic-name"),
		StringValue:   inventoryProductAddedTopic.TopicName(),
	})
}
